Sources/Event.swift

```diff
@@ -832,3 +837,98 @@ extension Signal.Event where Value == Never {
 		}
 	}
 }
+
+private enum EventStreamStatus: Int32 {
+	case alive
+	case terminated
+	case terminationBlocked
+}
+
+extension Signal.Event {
+	/// Wrap the given event sink with the necessary synchronization logic so
+	/// that it satisfies the `Signal` contract, specifically the serial event
+	/// order and the support for recursively delivered terminal events.
+	///
+	/// Recursive events are disallowed for `value` events, but are permitted
+	/// for termination events. Specifically:
+	///
+	/// - `interrupted`
+	/// It can inadvertently be sent by downstream consumers as part of the
+	/// `SignalProducer` mechanics.
+	///
+	/// - `completed`
+	/// If a downstream consumer weakly references an object, invocation of
+	/// such consumer may cause a race condition with its weak retain against
+	/// the last strong release of the object. If the `Lifetime` of the
+	/// object is being referenced by an upstream `take(during:)`, a
+	/// signal recursion might occur.
+	///
+	/// So we would treat termination events specially. If it happens to
+	/// occur while the `sendLock` is acquired, the observer call-out and
+	/// the disposal would be delegated to the current sender, or
+	/// occasionally one of the senders waiting on `sendLock`.
+	internal static func makeSynchronizing(_ action: @escaping Signal.Observer.Action, disposable: Disposable? = nil) -> Signal.Observer.Action {
+		let sendLock = Lock.make()
+		let status = UnsafeAtomicState(EventStreamStatus.alive)
+		let deallocator = ScopedDisposable(AnyDisposable(status.deinitialize))
+
+		// No one loves IUO, but logically speaking it is always available when
+		// stream state is `terminationBlocked`.
+		//
+		// This variable is only write once and subsequently read once.
+		var terminalEvent: Signal.Event!
+
+		return { [deallocator] event in
+			// All `sendLock` holders for delivering values must invoke
+			// `tryToCommitTermination` after releasing the lock. This ensures
+			// the terminal event would eventually be picked up.
+			@inline(__always)
+			func tryToCommitTermination() {
+				if status.is(.terminationBlocked) && sendLock.try() {
+					// The transition CAS here acts as an acquire fence for
+					// `terminalEvent`.
+					let shouldTerminate = status.tryTransition(from: .terminationBlocked, to: .terminated)
+
+					if shouldTerminate {
+						action(terminalEvent)
+						_ = deallocator
+					}
+
+					sendLock.unlock()
+
+					if shouldTerminate {
+						disposable?.dispose()
+					}
+				}
+			}
+
+			if case .value = event {
+				guard status.is(.alive) else { return }
+
+				sendLock.lock()
+				action(event)
+				sendLock.unlock()
+
+				tryToCommitTermination()
+			} else if status.tryTransition(from: .alive, to: .terminated) {
+				// Gracefully ignore any terminal event other than the one
+				// transitions the event stream away from the `alive` status.
+				guard !sendLock.try() else {
```

Comment:
I would invert this `guard`. I think it reads weirdly. The common case (unblocked termination) should be what we're guarding.